File: Competitive Programming/codeforces/training/F_Selling_a_Menagerie.py
# Copyright (c) 2023, Le Duc Phuc Long

# If you don't think twice, you have to code twice.

# Import session
import sys
from collections import defaultdict

############ ---- Input Functions ---- ############
def inp():
    return int(input())

def inlt():
    return list(map(int, input().split()))

def instr():
    return list(input().strip())

def invr():
    return map(int, input().split())

############ ---- Other Functions ---- ############
# Precompute
    
# Main function
# sys.stdin = open('in.txt', 'r')
for _ in range(inp()):
    # Input
    n = inp()
    a = inlt()
    b = inlt()
    # Count number of animals afraid of the animal a[i]
    cnt_afraid = [0]*n
    for i in range(n):
        a[i] -= 1
        cnt_afraid[a[i]] += 1
    # Put all animals don't afraid anyone into a list
    shelf = []
    for i in range(n):
        if not cnt_afraid[i]:
            shelf.append(i)
    # Sell all of those animals
    ans = [] # This list contains the order to sellings animals
    while shelf:
        sell_animal = shelf.pop()
        ans.append(sell_animal)
        cnt_afraid[a[sell_animal]] -= 1
        if cnt_afraid[a[sell_animal]] == 0:
            shelf.append(a[sell_animal])
    # Now we will iterate over all the cycles to add them to the ans list
    for u in range(n):
        if cnt_afraid[u]:
            # Find the cycle
            cycle = []
            v = a[u]
            while v != u:
                cycle.append([b[v], v])
                v = a[v]
            cycle.append([b[u], u])
            # Add it to ans optimally
            fin = min(cycle)[1]
            sta = a[fin]
            while sta != fin:
                ans.append(sta)
                cnt_afraid[sta] = 0
                sta = a[sta]
            ans.append(fin)
            cnt_afraid[fin] = 0
    print(*map(lambda x: x + 1, ans))
            
            
            
        
    
    

File: Competitive Programming/codeforces/training/F_Selling_a_Menagerie.cpp
// Copyright (c) 2023, Le Duc Phuc Long

/*If you don't think twice, you have to code twice.*/

// Include
#include <bits/stdc++.h>  // NOLINT
// #include <conio.h>

// Namespace
using namespace std;  // NOLINT

// Define
#define nln '\n'

// Typedef and Struct
typedef int64_t lo;

// Constant

// Declare global variables.
// End of global variable declaration.

void precompute() {
}

void solve() {
    // Input
    map<lo, lo> afraid;
    lo n;
    cin >> n;
    vector<lo> a(n+1);
    for (lo i = 1; i <= n; ++i) {
        cin >> a[i];
        afraid[i] = a[i];
    }
    vector<lo> c(n+1);
    for (lo i = 1; i <= n; ++i)
        cin >> c[i];
    // Solve
    // Sell all the animals that not afraid any other animal.
    for (lo i = 1; i <= n; ++i)
        if (afraid[i] == 0)
            cout << i << ' ';
           
}

int main(int argc, char* argv[]) {
    cin.tie(0)->sync_with_stdio(0);
    cout.tie(0)->sync_with_stdio(0);
    int T = 1;
    cin >> T;
    precompute();
    while (T--) {
        solve();
    }
    cerr << "It's ok Long, Good for now !!" << nln;
    return 0;
}